x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStoreTests.java

```diff
@@ -1978,6 +1981,149 @@ public void testApiKeyAuthUsesApiKeyServiceWithScopedRole() throws Exception {
+        }).when(nativePrivStore).getPrivileges(anyCollection(), anyCollection(), anyActionListener());
+
+        final AtomicReference<Collection<RoleDescriptor>> effectiveRoleDescriptors = new AtomicReference<Collection<RoleDescriptor>>();
+        final CompositeRolesStore compositeRolesStore = buildCompositeRolesStore(
+            SECURITY_ENABLED_SETTINGS,
+            fileRolesStore,
+            nativeRolesStore,
+            reservedRolesStore,
+            nativePrivStore,
+            null,
+            apiKeyService,
+            null,
+            null,
+            effectiveRoleDescriptors::set
+        );
+        AuditUtil.getOrGenerateRequestId(threadContext);
+        final TransportVersion version = TransportVersion.CURRENT;
+        final String apiKeyRoleName = "user_role_" + randomAlphaOfLength(4);
+        final Authentication apiKeyAuthentication = createApiKeyAuthentication(
+            apiKeyService,
+            randomValueOtherThanMany(
+                authc -> authc.getAuthenticationType() == AuthenticationType.API_KEY,
+                () -> AuthenticationTestHelper.builder().build()
+            ),
+            Collections.singleton(
+                new RoleDescriptor(
+                    apiKeyRoleName,
+                    null,
+                    new IndicesPrivileges[] { IndicesPrivileges.builder().indices("index*").privileges("all").build() },
+                    null
+                )
+            ),
+            null,
+            version
+        );
+        final String roleName = RBACEngine.REMOTE_USER_ROLE_NAME;
+        final boolean emptyRemoteRole = randomBoolean();
+        final Authentication authentication = apiKeyAuthentication.toRemoteAccess(
+            AuthenticationTestHelper.randomRemoteAccessAuthentication(
+                emptyRemoteRole
+                    ? RoleDescriptorsIntersection.EMPTY
+                    : new RoleDescriptorsIntersection(
+                        new RoleDescriptor(
+                            roleName,
+                            null,
+                            new RoleDescriptor.IndicesPrivileges[] {
+                                RoleDescriptor.IndicesPrivileges.builder().indices("index1").privileges("read").build() },
+                            null,
+                            null,
+                            null,
+                            null,
+                            null,
+                            null
+                        )
+                    )
+            )
+        );
+
+        final PlainActionFuture<Role> roleFuture = new PlainActionFuture<>();
+        compositeRolesStore.getRole(authentication.getEffectiveSubject(), roleFuture);
+        final Role role = roleFuture.actionGet();
+        assertThat(effectiveRoleDescriptors.get(), is(nullValue()));
+
+        verify(apiKeyService, times(1)).parseRoleDescriptorsBytes(anyString(), any(BytesReference.class), any());
+        if (emptyRemoteRole) {
```

Comment:
It's quirky behavior in both situations: when the FC receives an empty role descriptor intersection, we have no role names (even though the remote access API key has some), when the intersection is non-empty the role name is always the same, regardless of querying cluster or querying subject (it's hard-coded to `_remote_user` in RBACEngine). 

This is partly due to the limitation we have in `LimitedRole.names()`. We could potentially improve by addressing the current limitation of `LimitedRole.names()` and by fetching role names from the nested authentication instance stored in metadata. The latter is likely less work but I would still like to defer it: the primary benefit of having more complete role info would come in when we deal with auditing and potentially crafting RCS 2.0 specific authz denied messages, which are issues prioritized for later down the line.

Reply:
I think an empty role names is the right thing for subjects that do not have roles but just privileges. We have this concept but it is not always the case in the code, e.g. `user.roles` in audit logging. It is mostly for historical reasons and I don't feel we have to follow. 

This is also related to my above suggestion about changing the order of roleReference list. If we make the change, the role names will always be the API key's owner user's roles. This is more consistent (in that it is the same result) but likely wrong because the corresponding `user.name` field in audit logging is of the remote user.

In any case, we can address it later in the audit logging work.